source/portable/NetworkInterface/xilinx_ultrascale/NetworkInterface.c

```diff
@@ -123,116 +126,259 @@ static void prvEMACHandlerTask( void * pvParameters );
-            }
+            #if ( USE_JUMBO_FRAMES == 1 )
+                /* Enable jumbo frames for zynqmp */
+                XEmacPs_SetOptions( pxEMAC_PS, XEMACPS_JUMBO_ENABLE_OPTION );
+            #endif
 
-            /* Initialize the mac and set the MAC address. */
-            XEmacPs_SetMacAddress( pxEMAC_PS, ( void * ) ipLOCAL_MAC_ADDRESS, 1 );
+            /* Initialize the mac and set the MAC address at position 1. */
+            XEmacPs_SetMacAddress( pxEMAC_PS, ( void * ) pxEndPoint->xMACAddress.ucBytes, 1 );
 
             #if ( ipconfigUSE_LLMNR == 1 )
             {
-                /* Also add LLMNR multicast MAC address. */
-                XEmacPs_SetMacAddress( pxEMAC_PS, ( void * ) xLLMNR_MACAddress, 2 );
+                #if ( ipconfigUSE_IPv6 == 0 )
+                {
+                    /* Also add LLMNR multicast MAC address. */
+                    XEmacPs_SetHash( pxEMAC_PS, ( void * ) xLLMNR_MacAddress.ucBytes );
+                }
+                #else
+                {
+                    /* is this loop really required? */
+                    for( pxEndPoint = FreeRTOS_FirstEndPoint( pxInterface );
+                         pxEndPoint != NULL;
+                         pxEndPoint = FreeRTOS_NextEndPoint( pxInterface, pxEndPoint ) )
+                    {
+                        if( pxEndPoint->bits.bIPv6 != pdFALSE_UNSIGNED )
+                        {
+                            unsigned char ucMACAddress[ 6 ] = { 0x33, 0x33, 0xff, 0, 0, 0 };
+                            ucMACAddress[ 3 ] = pxEndPoint->ipv6_settings.xIPAddress.ucBytes[ 13 ];
+                            ucMACAddress[ 4 ] = pxEndPoint->ipv6_settings.xIPAddress.ucBytes[ 14 ];
+                            ucMACAddress[ 5 ] = pxEndPoint->ipv6_settings.xIPAddress.ucBytes[ 15 ];
+                            XEmacPs_SetHash( pxEMAC_PS, ( void * ) ucMACAddress );
+                        }
+                    }
+
+                    XEmacPs_SetHash( pxEMAC_PS, ( void * ) xLLMNR_MacAddressIPv6.ucBytes );
+                }
+                #endif /* if ( ipconfigUSE_IPv6 == 0 ) */
+
+
+                /* allow reception of multicast addresses programmed into hash */
+                XEmacPs_SetOptions( pxEMAC_PS, XEMACPS_MULTICAST_OPTION );
             }
             #endif /* ipconfigUSE_LLMNR == 1 */
 
+            #if ( ( ipconfigUSE_MDNS == 1 ) && ( ipconfigUSE_IPv6 != 0 ) )
+                XEmacPs_SetHash( pxEMAC_PS, ( void * ) xMDNS_MacAddress.ucBytes ); /* why for IPv6? */
+                XEmacPs_SetHash( pxEMAC_PS, ( void * ) xMDNS_MACAddressIPv6.ucBytes );
+
+                /* allow reception of multicast addresses programmed into hash */
+                XEmacPs_SetOptions( pxEMAC_PS, XEMACPS_MULTICAST_OPTION );
+            #endif /* ( ( ipconfigUSE_MDNS == 1 ) && ( ipconfigUSE_IPv6 != 0 ) ) */
+
+            /* In the Zynq7000 port the MAC-Address is stored on second endpoint here at position 4. Needed? */
```

Comment:
As mentioned in the comment Zynq7000 does this here:
```
        pxEndPoint = FreeRTOS_NextEndPoint( pxInterface, pxEndPoint );

        if( pxEndPoint != NULL )
        {
            /* If there is a second end-point, store the MAC
             * address at position 4.*/
            XEmacPs_SetMacAddress( pxEMAC_PS, ( void * ) pxEndPoint->xMACAddress.ucBytes, 4 );
        }
```
Do we also need this here?